lib/utils.hpp

```diff
@@ -0,0 +1,34 @@
+// Utilities in the system domain (like logging)
+
+#pragma once
+
+#include <iostream>
+#include <limits>
+#include <string>
+#include <vector>
+
+// Log a string to console
+inline void log(std::string message) {
+    std::cout << message << "\n";
+}
+
+// Log a vector<double> to console
+inline void log(std::vector<double> row) {
+  // log numbers with full precision, usually 17 digits
+  std::cout.precision(std::numeric_limits<double>::max_digits10);
+
+  std::cout << "{";
+  for (double val : row) {
+    std::cout << val << ", ";
+  }
+  std::cout << "}"
+       << "\n";
+}
+
+// Throw exception if index is out of bounds of vector
+template <class T>
```

Comment:
C++ obeys the separate compilation principle: a template function in utils.cpp must know at compilation time what classes it will be called for, otherwise it will fail to link against specialized invocations. 

To ensure vectorGuard will compile against any new vector types in the future, I've moved it into the header.

# Example

This fails:  
a.hpp:
```c++
#include <vector>

template <class T>
void myFunc(std::vector<T> a);
```
a.cpp
```c++
#include <vector>
#include "a.hpp"

template  <class T>
void myFunc(std::vector<T> a){}
```
b.cpp
```c++
#include <vector>
#include "a.hpp"

int main() {
  std::vector<int> x;
  myFunc(x);
}
```
Result:
```
b.cpp: undefined reference to `void myFunc<int>(std::vector<int, std::allocator<int> >)'
```